meson.build

```diff
@@ -0,0 +1,717 @@
+cdata.set('GEANY_DATA_DIR', join_paths(prefix, get_option('datadir'), 'geany'))
+cdata.set('GEANY_DOC_DIR', join_paths(prefix, get_option('datadir'), 'doc', 'geany'))
+cdata.set('top_srcdir', meson.source_root())
+cdata.set('top_builddir', meson.build_root())
+
+# for geany.pc (adapted from GTK+)
+pcconf = cdata
+pcconf.set('DEPENDENCIES', deps_for_pc)
+pcconf.set('prefix', get_option('prefix'))
+pcconf.set('libdir', join_paths('${exec_prefix}', get_option('libdir')))
+pcconf.set('includedir', join_paths('${prefix}', get_option('includedir')))
+pcconf.set('datarootdir', join_paths('${prefix}', get_option('datadir')))
+# actually constant, just match autconf
+pcconf.set('exec_prefix', '${prefix}')
+pcconf.set('datadir', '${datarootdir}')
+pcconf.set('localedir', join_paths('${datarootdir}', 'locale'))
+
+# needed programs
+sh = find_program('sh')
+cp = find_program('cp')
+ln = find_program('ln')
+# python is not strictly required but present anyway (due to meson)
+python = find_program('python3', 'python')
+# These two are truly optional
+rst2html = find_program('rst2html', required: false)
+rst2pdf = find_program('rst2pdf', required: false)
+git = find_program('git', required: false)
+if git.found()
+	ret = run_command(git, 'rev-parse', '--short', '--revs-only', 'HEAD')
+	cdata.set_quoted('REVISION', ret.stdout().strip())
+else
+	cdata.set_quoted('REVISION', '-1')
+endif
+
+configure_file(input : 'config.h.meson',
+  output : 'config.h',
+  configuration : cdata)
+
+configure_file(input: 'geany.pc.in',
+               install: true,
+               install_dir: join_paths(prefix, get_option('libdir'), 'pkgconfig'),
+               output: 'geany.pc',
+               configuration: pcconf)
+
+# CFLAGS for basic stuff that only depends on libc
+basic_cflags  = [ '-DHAVE_CONFIG_H=1', '-O2' ]
+if cdata.get('HAVE_GNU_GET_LIBC_VERSION')
+    basic_cflags += '-D_GNU_SOURCE'
+endif
+
+# CFLAGS common between Geany and bundled plugins
+def_cflags = basic_cflags + [ '-DGTK' ]
+# we're using lots of deprecated stuff with gtk3 (often because there is no proper alternative!)
+def_cflags += '-Wno-deprecated-declarations'
+def_cflags += '-DGDK_DISABLE_DEPRECATION_WARNINGS'
+def_cflags += '-DGLIB_VERSION_MIN_REQUIRED=GLIB_VERSION_2_32'
+def_cflags += '-DGEANY_PREFIX="@0@"'.format(get_option('prefix'))
+foreach d : [ 'includedir', 'libdir', 'libexecdir', 'datadir', 'localedir' ]
+  def_cflags += '-DGEANY_@0@="@1@"'.format(d.underscorify().to_upper(), join_paths(prefix, get_option(d)))
+endforeach
+def_cflags += '-DGEANY_DOCDIR="@0@"'.format(join_paths(prefix, get_option('datadir'), 'doc'))
+
+# CFLAGS for everything else, i.e. most of Geany
+geany_cflags = def_cflags
+have_gcc4_visibility = cc.has_argument('-fvisibility=hidden')
+geany_cflags += '-DGEANY_PRIVATE'
+if target_machine.system() == 'windows'
+  geany_cflags += '-DGEANY_EXPORT_SYMBOL="__declspec(dllexport)"'
+elif have_gcc4_visibility
+  geany_cflags += '-fvisibility=hidden'
```

Comment:
Use the meson kwarg `gnu_symbol_visibility: 'hidden'` on your build targets, and let the declspec define fill in the support gap for Windows (I think everywhere else should support it, really).